File: frontend/src/app/common/entities/skill.ts
import { range } from 'rxjs';
import { Status } from '../control/status';
import { Position } from '../geometry/position';
export class Skill {
  points: number = 0;
  bonusPoints: number = 0;
  status: Status = new Status();
  cooldownTimer = 0;
  usages: number = 0;
  maxPoints: number = 0;
  chakraCost: number = 0;
  cooldown: number = 0;
  castTime: number = 0;
  dmgBase: number = 0;
  cScale: number = 0;
  cNumber: number = 0;
  mMaxHealthBase: number = 0;
  bMaxHealth: number = 0;
  bMaxChakra: number = 0;
  bTaijutsuPower: number = 0;
  bNinjutsuPower: number = 0;
  bAttackSpeed: number = 0;
  bMovementSpeed: number = 0;
  bResistances: number = 0;
  bSideEffectTaijutsuPower: number = 0;
  bSideEffectNinjutsuPower: number = 0;
  bSideEffectAttackSpeed: number = 0;
  bSideEffectMovementSpeed: number = 0;
  bSideEffectResistances: number = 0;
  dResistances: number = 0;
  aArmorBase = 0;
  pMaxHealth: number = 0;
  pMaxChakra: number = 0;
  pTaijutsuPower: number = 0;
  pNinjutsuPower: number = 0;
  pAttackSpeed: number = 0;
  pMovementSpeed: number = 0;
  pResistances: number = 0;
  hStacksBonus: number = 0;
  constructor(
    public id: number = -1,
    public name: string = '',
    public imageUrl: string = '',
    public tier: number = -1,
    public target: string = 'none', //'enemy, self, tile, none'
    public p_chakraCost: number = 0,
    public p_chakraCostPerPoint: number = 0,
    public p_cooldown: number = 0,
    public p_cooldownPerPoint: number = 0,
    public p_castTime: number = 0,
    public p_castTimePerPoint: number = 0,
    public range: number = 0,
    public dmgType: string = '',
    public p_dmgBase: number = 0,
    public p_dmgBasePerPoint: number = 0,
    public dmgTaijutsuPowerScale: number = 0,
    public dmgNinjutsuPowerScale: number = 0,
    public dmgMaxHealthScale: number = 0,
    public dmgMaxChakraScale: number = 0,
    public dmgIncreasePerDistanceCoveredPerc = 0,
    public dmgIncreasePerMissingHealthPerc = 0,
    public move: boolean = false,
    public clone: boolean = false,
    public c_weak: boolean = false,
    public c_swap: boolean = false,
    public p_cScale: number = 0,
    public p_cScalePerPoint: number = 0,
    public p_cNumber: number = 0,
    public p_cNumberPerPoint: number = 0,
    public mirror: boolean = false,
    public p_mMaxHealthBase: number = 0,
    public p_mMaxHealthBasePerPoint: number = 0,
    public m_ninjutsuScale: number = 0,
    public invisibility: boolean = false,
    public buff: boolean = false,
    public p_bMaxHealth: number = 0,
    public p_bMaxChakra: number = 0,
    public p_bTaijutsuPower: number = 0,
    public p_bNinjutsuPower: number = 0,
    public p_bAttackSpeed: number = 0,
    public p_bMovementSpeed: number = 0,
    public p_bResistances: number = 0,
    public b_sideEffect: boolean = false,
    public p_bSideEffectTaijutsuPower: number = 0,
    public p_bSideEffectNinjutsuPower: number = 0,
    public p_bSideEffectAttackSpeed: number = 0,
    public p_bSideEffectMovementSpeed: number = 0,
    public p_bSideEffectResistances: number = 0,
    public p_bMaxHealthPerPoint: number = 0,
    public p_bMaxChakraPerPoint: number = 0,
    public p_bTaijutsuPowerPerPoint: number = 0,
    public p_bNinjutsuPowerPerPoint: number = 0,
    public p_bAttackSpeedPerPoint: number = 0,
    public p_bMovementSpeedPerPoint: number = 0,
    public p_bResistancesPerPoint: number = 0,
    public p_bSideEffectTaijutsuPowerPerPoint: number = 0,
    public p_bSideEffectNinjutsuPowerPerPoint: number = 0,
    public p_bSideEffectAttackSpeedPerPoint: number = 0,
    public p_bSideEffectMovementSpeedPerPoint: number = 0,
    public p_bSideEffectResistancesPerPoint: number = 0,
    public kyuubiChakra: boolean = false,
    public requiresKyuubiChakra: boolean = false,
    public throwInAir: boolean = false,
    public requiresTargetInAir: boolean = false,
    public debuff = false,
    public p_dResistances: number = 0,
    public p_dResistancesPerPoint: number = 0,
    public armor = false,
    public p_aArmorBase = 0,
    public p_aArmorBasePerPoint = 0,
    public a_ninjutsuScale = 0,
    public passive: boolean = false,
    public p_pMaxHealth: number = 0,
    public p_pMaxChakra: number = 0,
    public p_pTaijutsuPower: number = 0,
    public p_pNinjutsuPower: number = 0,
    public p_pAttackSpeed: number = 0,
    public p_pMovementSpeed: number = 0,
    public p_pResistances: number = 0,
    public p_pMaxHealthPerPoint: number = 0,
    public p_pMaxChakraPerPoint: number = 0,
    public p_pTaijutsuPowerPerPoint: number = 0,
    public p_pNinjutsuPowerPerPoint: number = 0,
    public p_pAttackSpeedPerPoint: number = 0,
    public p_pMovementSpeedPerPoint: number = 0,
    public p_pResistancesPerPoint: number = 0,
    public hyuugaStyle: boolean = false,
    public p_hStacksBonus: number = 0,
    public p_hStacksBonusPerPoint: number = 0
  ) {}
  update() {
    if (this.tier === 1) {
      this.maxPoints = 5;
    }
    if (this.tier === 2) {
      this.maxPoints = 5;
    }
    if (this.tier === 3) {
      this.maxPoints = 3;
    }
    if (this.points > 1) {
      this.bonusPoints = this.points - 1;
    } else {
      this.bonusPoints = 0;
    }
    this.chakraCost =
      this.bonusPoints * this.p_chakraCostPerPoint + this.p_chakraCost;
    this.cooldown =
      this.bonusPoints * this.p_cooldownPerPoint + this.p_cooldown;
    this.castTime =
      this.bonusPoints * this.p_castTimePerPoint + this.p_castTime;
    this.dmgBase = this.bonusPoints * this.p_dmgBasePerPoint + this.p_dmgBase;
    this.cScale = this.bonusPoints * this.p_cScalePerPoint + this.p_cScale;
    this.cNumber = this.bonusPoints * this.p_cNumberPerPoint + this.p_cNumber;
    this.mMaxHealthBase =
      this.bonusPoints * this.p_mMaxHealthBasePerPoint + this.p_mMaxHealthBase;
    this.bMaxHealth =
      this.bonusPoints * this.p_bMaxHealthPerPoint + this.p_bMaxHealth;
    this.bMaxChakra =
      this.bonusPoints * this.p_bMaxChakraPerPoint + this.p_bMaxChakra;
    this.bTaijutsuPower =
      this.bonusPoints * this.p_bTaijutsuPowerPerPoint + this.p_bTaijutsuPower;
    this.bNinjutsuPower =
      this.bonusPoints * this.p_bNinjutsuPowerPerPoint + this.p_bNinjutsuPower;
    this.bAttackSpeed =
      this.bonusPoints * this.p_bAttackSpeedPerPoint + this.p_bAttackSpeed;
    this.bMovementSpeed =
      this.bonusPoints * this.p_bMovementSpeedPerPoint + this.p_bMovementSpeed;
    this.bResistances =
      this.bonusPoints * this.p_bResistancesPerPoint + this.p_bResistances;
    this.bSideEffectTaijutsuPower =
      this.bonusPoints * this.p_bSideEffectTaijutsuPowerPerPoint +
      this.p_bSideEffectTaijutsuPower;
    this.bSideEffectNinjutsuPower =
      this.bonusPoints * this.p_bSideEffectNinjutsuPowerPerPoint +
      this.p_bSideEffectNinjutsuPower;
    this.bSideEffectAttackSpeed =
      this.bonusPoints * this.p_bSideEffectAttackSpeedPerPoint +
      this.p_bSideEffectAttackSpeed;
    this.bSideEffectMovementSpeed =
      this.bonusPoints * this.p_bSideEffectMovementSpeedPerPoint +
      this.p_bSideEffectMovementSpeed;
    this.bSideEffectResistances =
      this.bonusPoints * this.p_bSideEffectResistancesPerPoint +
      this.p_bSideEffectResistances;
    this.dResistances =
      this.bonusPoints * this.p_dResistancesPerPoint + this.p_dResistances;
    this.aArmorBase =
      this.bonusPoints * this.p_aArmorBasePerPoint + this.p_aArmorBase;
    this.pMaxHealth =
      this.bonusPoints * this.p_pMaxHealthPerPoint + this.p_pMaxHealth;
    this.pMaxChakra =
      this.bonusPoints * this.p_pMaxChakraPerPoint + this.p_pMaxChakra;
    this.pTaijutsuPower =
      this.bonusPoints * this.p_pTaijutsuPowerPerPoint + this.p_pTaijutsuPower;
    this.pNinjutsuPower =
      this.bonusPoints * this.p_pNinjutsuPowerPerPoint + this.p_pNinjutsuPower;
    this.pAttackSpeed =
      this.bonusPoints * this.p_pAttackSpeedPerPoint + this.p_pAttackSpeed;
    this.pMovementSpeed =
      this.bonusPoints * this.p_pMovementSpeedPerPoint + this.p_pMovementSpeed;
    this.pResistances =
      this.bonusPoints * this.p_pResistancesPerPoint + this.p_pResistances;
    this.hStacksBonus =
      this.bonusPoints * this.p_hStacksBonusPerPoint + this.p_hStacksBonus;
  }
  startCooldownTimer() {
    this.cooldownTimer = this.cooldown;
  }
  cooldownTimerDecrement() {
    if (this.cooldownTimer > 0) {
      this.cooldownTimer--;
    }
  }
}

File: backend/src/main/java/com/jozwiaka/backend/entity/Skill.java
package com.jozwiaka.backend.entity;

import jakarta.persistence.*;
import lombok.Data;

import java.io.Serializable;

@Entity
@Data
@Table(name = "skill")
public class Skill implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "tree_id")
    private Tree tree;

    @Column(unique = true, nullable = false)
    private String name;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(name = "tier")
    private Double tier;

    @Column(name = "target")
    private String target;

    @Column(name = "p_chakra_cost")
    private Double p_chakraCost;

    @Column(name = "p_chakra_cost_per_point")
    private Double p_chakraCostPerPoint;

    @Column(name = "p_cooldown")
    private Double p_cooldown;

    @Column(name = "p_cooldown_per_point")
    private Double p_cooldownPerPoint;

    @Column(name = "p_cast_time")
    private Double p_castTime;

    @Column(name = "p_cast_time_per_point")
    private Double p_castTimePerPoint;

    @Column(name = "range")
    private Double range;

    @Column(name = "dmg_type")
    private String dmgType;

    @Column(name = "p_dmg_base")
    private Double p_dmgBase;

    @Column(name = "p_dmg_base_per_point")
    private Double p_dmgBasePerPoint;

    @Column(name = "dmg_taijutsu_power_scale")
    private Double dmgTaijutsuPowerScale;

    @Column(name = "dmg_ninjutsu_power_scale")
    private Double dmgNinjutsuPowerScale;

    @Column(name = "dmg_max_health_scale")
    private Double dmgMaxHealthScale;

    @Column(name = "dmg_max_chakra_scale")
    private Double dmgMaxChakraScale;

    @Column(name = "dmg_increase_per_distance_covered_perc")
    private Double dmgIncreasePerDistanceCoveredPerc;

    @Column(name = "dmg_increase_per_missing_health_perc")
    private Double dmgIncreasePerMissingHealthPerc;

    @Column(name = "move")
    private boolean move;

    @Column(name = "clone")
    private boolean clone;

    @Column(name = "c_weak")
    private boolean c_weak;

    @Column(name = "c_swap")
    private boolean c_swap;

    @Column(name = "p_c_scale")
    private Double p_cScale;

    @Column(name = "p_c_scale_per_point")
    private Double p_cScalePerPoint;

    @Column(name = "p_c_number")
    private Double p_cNumber;

    @Column(name = "p_c_number_per_point")
    private Double p_cNumberPerPoint;

    @Column(name = "mirror")
    private boolean mirror;

    @Column(name = "p_m_max_health_base")
    private Double p_mMaxHealthBase;

    @Column(name = "p_m_max_health_base_per_point")
    private Double p_mMaxHealthBasePerPoint;

    @Column(name = "m_ninjutsu_scale")
    private Double m_ninjutsuScale;

    @Column(name = "invisibility")
    private boolean invisibility;

    @Column(name = "buff")
    private boolean buff;

    @Column(name = "p_b_max_health")
    private Double p_bMaxHealth;

    @Column(name = "p_b_max_chakra")
    private Double p_bMaxChakra;

    @Column(name = "p_b_taijutsu_power")
    private Double p_bTaijutsuPower;

    @Column(name = "p_b_ninjutsu_power")
    private Double p_bNinjutsuPower;

    @Column(name = "p_b_attack_speed")
    private Double p_bAttackSpeed;

    @Column(name = "p_b_movement_speed")
    private Double p_bMovementSpeed;

    @Column(name = "p_b_resistances")
    private Double p_bResistances;

    @Column(name = "b_side_effect")
    private boolean b_sideEffect;

    @Column(name = "p_b_side_effect_taijutsu_power")
    private Double p_bSideEffectTaijutsuPower;

    @Column(name = "p_b_side_effect_ninjutsu_power")
    private Double p_bSideEffectNinjutsuPower;

    @Column(name = "p_b_side_effect_attack_speed")
    private Double p_bSideEffectAttackSpeed;

    @Column(name = "p_b_side_effect_movement_speed")
    private Double p_bSideEffectMovementSpeed;

    @Column(name = "p_b_side_effect_resistances")
    private Double p_bSideEffectResistances;

    @Column(name = "p_b_max_health_per_point")
    private Double p_bMaxHealthPerPoint;

    @Column(name = "p_b_max_chakra_per_point")
    private Double p_bMaxChakraPerPoint;

    @Column(name = "p_b_taijutsu_power_per_point")
    private Double p_bTaijutsuPowerPerPoint;

    @Column(name = "p_b_ninjutsu_power_per_point")
    private Double p_bNinjutsuPowerPerPoint;

    @Column(name = "p_b_attack_speed_per_point")
    private Double p_bAttackSpeedPerPoint;

    @Column(name = "p_b_movement_speed_per_point")
    private Double p_bMovementSpeedPerPoint;

    @Column(name = "p_b_resistances_per_point")
    private Double p_bResistancesPerPoint;

    @Column(name = "p_b_side_effect_taijutsu_power_per_point")
    private Double p_bSideEffectTaijutsuPowerPerPoint;

    @Column(name = "p_b_side_effect_ninjutsu_power_per_point")
    private Double p_bSideEffectNinjutsuPowerPerPoint;

    @Column(name = "p_b_side_effect_attack_speed_per_point")
    private Double p_bSideEffectAttackSpeedPerPoint;

    @Column(name = "p_b_side_effect_movement_speed_per_point")
    private Double p_bSideEffectMovementSpeedPerPoint;

    @Column(name = "p_b_side_effect_resistances_per_point")
    private Double p_bSideEffectResistancesPerPoint;

    @Column(name = "kyuubi_chakra")
    private boolean kyuubiChakra;

    @Column(name = "requires_kyuubi_chakra")
    private boolean requiresKyuubiChakra;

    @Column(name = "throw_in_air")
    private boolean throwInAir;

    @Column(name = "requires_target_in_air")
    private boolean requiresTargetInAir;

    @Column(name = "debuff")
    private boolean debuff;

    @Column(name = "p_d_resistances")
    private Double p_dResistances;

    @Column(name = "p_d_resistances_per_point")
    private Double p_dResistancesPerPoint;

    @Column(name = "armor")
    private boolean armor;

    @Column(name = "p_a_armor_base")
    private Double p_aArmorBase;

    @Column(name = "p_a_armor_base_per_point")
    private Double p_aArmorBasePerPoint;

    @Column(name = "a_ninjutsu_scale")
    private Double a_ninjutsuScale;

    @Column(name = "passive")
    private boolean passive;

    @Column(name = "p_p_max_health")
    private Double p_pMaxHealth;

    @Column(name = "p_p_max_chakra")
    private Double p_pMaxChakra;

    @Column(name = "p_p_taijutsu_power")
    private Double p_pTaijutsuPower;

    @Column(name = "p_p_ninjutsu_power")
    private Double p_pNinjutsuPower;

    @Column(name = "p_p_attack_speed")
    private Double p_pAttackSpeed;

    @Column(name = "p_p_movement_speed")
    private Double p_pMovementSpeed;

    @Column(name = "p_p_resistances")
    private Double p_pResistances;

    @Column(name = "p_p_max_health_per_point")
    private Double p_pMaxHealthPerPoint;

    @Column(name = "p_p_max_chakra_per_point")
    private Double p_pMaxChakraPerPoint;

    @Column(name = "p_p_taijutsu_power_per_point")
    private Double p_pTaijutsuPowerPerPoint;

    @Column(name = "p_p_ninjutsu_power_per_point")
    private Double p_pNinjutsuPowerPerPoint;

    @Column(name = "p_p_attack_speed_per_point")
    private Double p_pAttackSpeedPerPoint;

    @Column(name = "p_p_movement_speed_per_point")
    private Double p_pMovementSpeedPerPoint;

    @Column(name = "p_p_resistances_per_point")
    private Double p_pResistancesPerPoint;

    @Column(name = "hyuuga_style")
    private boolean hyuugaStyle;

    @Column(name = "p_h_stacks_bonus")
    private Double p_hStacksBonus;

    @Column(name = "p_h_stacks_bonus_per_point")
    private Double p_hStacksBonusPerPoint;
}
